plugins/org.eclipse.uml2.uml/src/org/eclipse/uml2/uml/util/UMLUtil.java

```diff
@@ -12432,15 +12972,47 @@ public static void setBaseElement(EObject stereotypeApplication,
 		}
 	}
 
+	private static EObject storeStereotypeApplication(Element element,
+			EClass definition, Stereotype stereotype) {
+		StereotypeApplicationStorage storage = UMLUtil
+			.isStereotypeApplicationStorageEnabled()
+				? UMLUtil.getStereotypeApplicationStorage(element)
+				: null;
+		if (storage != null) {
+			EObject stereotypeApplication = EcoreUtil.create(definition);
+			CacheAdapter.getInstance().adapt(stereotypeApplication);
+
+			storage.addStereotypeApplication(element, stereotype,
+				stereotypeApplication);
+
+			basicSetBaseElement(stereotypeApplication, definition, element);
+			return stereotypeApplication;
+		}
+
+		return null;
+	}
+
 	protected static EObject applyStereotype(Element element, EClass definition,
 			Stereotype stereotype) {
+		EObject stereotypeApplication = storeStereotypeApplication(element,
+			definition, stereotype);
+		if (stereotypeApplication != null) {
```

Comment:
This amounts to doing a Boolean check and two tests for `null` each time a stereotype is applied in cases where a custom storage is not being used (the default). Could we please optimize for the least overhead possible in the default case?

Reply:
Sorry for being unclear. I am suggesting moving the Boolean check to the beginning of the two `applyStereotype` methods and only calling the `storeStereotypeApplication` method if storage is enabled; otherwise fall through to the default (previous) logic. Doing so would avoid the two subsequent `null` checks (and extra method call) in cases where the default behaviour is being used.

Reply:
Ok, I've amended my commit again. It should be ready for your review now.